### PR TITLE
Fix race condition for concurrent `/auth/refresh` calls

### LINKED ISSUE #22360: Race condition on refresh with multiple tabs
### Describe the Bug

Currently the /refresh endpoint is called whenever a tab becomes active. There is an issue where this sometimes gets called more than once from different tabs and causes a logout to occur.

I'm not sure if this can be fixed with a time delay or a retry. 

### To Reproduce

open 2 Directus tabs in the same browser. Make sure both tabs are alongside each other.
Move one tab out and notice that one of the directus projects will be logged out.

https://github.com/directus/directus/assets/21194068/7fe70730-c50d-4c36-b53a-9a86e2d93150

### Directus Version

10.10.7

### Hosting Strategy

Self-Hosted (Docker Image)

### PR DESCRIPTION
**Pull request by hanneskuettner**
## Scope

In #22360 we uncovered that concurrent requests to the `/auth/refresh` endpoint end up causing a race condition where the user get's logged out if following happens:

**Scenario 1:**

- Window A requests `/auth/refresh` with a valid session cookie and session identifier (which is called `refreshToken` in the code) `token_1`
- Window B requests `/auth/refresh` with a the same session cookie and session identifier `token_1`
- Window A passes the valid session check since `token_1` is in the `directus_sessions` table
- Window A receives a new token `token_2`, updates that in the `directus_sessions` table, and receives a new session cookie with `session: token_2`
- Window B does not pass valid session check since `token_1` is not in the `directus_sessions` table anymore
- Window B gets logged out

**Scenario 2:**

Similar two Scenario 1, but the verification are more interleaved due to the async nature of the database / auth provider / hook calls we're making.

- Window A & B both pass the valid session check since `token_1` is in the `directus_sessions` table
- Window A updates the token to `token_2` and sets the session cookie
- Window B updates the token to `token_3` and sets the session cookie
- Window A does another request with while the session cookie from Window B is not yet received, and is being logged out because it still has the now invalid `token_2` set.

--- 

Since we're using the `refreshToken` as a session identifier, wrapped in the session cookie, in the `session` mode, it should safe to not update it in the DB. The signed session cookie with the expiry time coded in takes care of invalidating the session after `SESSION_COOKIE_TTL` and `REFRESH_TOKEN_TTL` takes care of invalidating the session stored in the DB. 
Not rotating the `refreshToken` (really the session identifier) should not impact the security as far as I 

What's changed:

- Don't refresh the session identifying token when using `session` mode

## Potential Risks / Drawbacks

Should be none

## Review Notes / Questions

- Is not rotating the session identifier a problem I can't think of?

---

Fixes #22360

### CODE REVIEW

**Review by br41nslug — CHANGES_REQUESTED** [review]
Since this PR prevents the session id (aka refresh token) from changing for session mode authentication we should prevent people from extracting the "session id / refresh token" from the session JWT and refreshing that like a regular `json` refresh token

image
image
image

As this will result in the same "invalid token" situation as the race condition currently does

**Comment by paescuj** [comment]
> Since this PR prevents the session id (aka refresh token) from changing for session mode authentication we should prevent people from extracting the "session id / refresh token" from the session JWT and refreshing that like a regular `json` refresh token

Could you help me understand why this would be necessary? If a user should do so, they certainly know what they are doing.

File: .changeset/hot-apples-greet.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+'@directus/api': patch
+---
+
+Fixed a race condition in `/auth/refresh` that occured when multiple app windows refresh at the same time
```
Comment (paescuj) [inline .changeset/hot-apples-greet.md:5]:
```suggestion
Fixed a race condition that occurred when multiple browser tabs refresh at the same time,  whereby a user could unintentionally lose the session
```

**Comment by br41nslug** [comment]
> Could you help me understand why this would be necessary?

One being functional, this is an "unintended feature" refreshing as session token as `json` breaks the session potentially leading to unexpected / unwanted results like getting logged out everywhere if using SSO. Imo if someone wants to use `json/cookie` mode you should log in using that mode not try to convert the stateful session token to a stateless access/refresh token set.
The purpose of this change is for the session ID to not change when using a stateful session and having a "workaround" that does explicitly change this undermines this fix.

So from a functional / footgun pov I believe we should be intentional with authentication and only be allowing this if there is a valid use-case to support that which outweighs the potential risks.

If this is a feature we want to support I'd be more in favor of intentionally implementing this in the login endpoint returning new credentials using a valid access_token and add proper documentation that this is possible.

> If a user should do so, they certainly know what they are doing.

True but we also have to consider malicious actors, while this is not directly exploitable out of the box however having an unused / undocumented feature in authentication inherently adds risk of abuse.

A basic example would be if any token during the session gets leaked (expired or not) can then be used to create a new session, locking out the current user and persisting access to the platform indefinitely.

**Review by paescuj — COMMENTED** [review]
Looks good to me so far:

- I'd even say it kinda makes more sense to not change the session identifier
- The session is still invalidated if the user logs-out
- As you mentioned, the session/refresh token will still expire when the user doesn't refresh

However, to @br41nslug's point, I agree we should make it more secure, by preventing usage of the session refresh token via other modes, ensuring the JWT expiration (`SESSION_COOKIE_TTL`) still takes place 👍 
This would probably mean to add a new field (`type` or similar) to the `directus_sessions` collection, which we can then compare against. This would go well with the middleware changes in #22327.

**Comment by br41nslug** [comment]
> This would probably mean to add a new field (type or similar) to the directus_sessions collection

I was considering something in that direction too 🤔 as i dont see many other options of telling if the specific session getting refreshed was intended to be stateless or stateful.

**Comment by paescuj** [comment]
> > This would probably mean to add a new field (type or similar) to the directus_sessions collection
> 
> I was considering something in that direction too 🤔 as i dont see many other options of telling if the specific session getting refreshed was intended to be stateless or stateful.

Yeah, another option could be to simply prefix session refresh token by something like `s_` but then we would have to make sure the other generated tokens never receive that prefix or additionally compare the length of the token, but this feels a bit hacky.

**Comment by br41nslug** [comment]
Interesting idea 🤔 that does indeed feel risky as the tokens are auto-generated and could theoretically generate one with that exact prefix (or any other prefix that consists of characters that are used by the nanoid generator)

**Comment by br41nslug** [comment]
After some discussion it looks like this frontend solution https://github.com/directus/directus/pull/22457 will take priority over the authentication change done here.

**Pull request closed without merging**